File: models/statement.py
"""Bank statement records."""

from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel
from sqlmodel import Field, SQLModel


def utc_now() -> datetime:
    from datetime import timezone

    return datetime.now(timezone.utc).replace(tzinfo=None)


class Statement(SQLModel, table=True):
    __tablename__ = "statements"  # type: ignore

    id: int | None = Field(default=None, primary_key=True)
    account_id: int = Field(foreign_key="accounts.id", index=True)
    period_start: date | None = None
    period_end: date | None = None
    previous_balance: float | None = None
    closing_balance: float | None = None
    minimum_payment: float | None = None
    due_date: date | None = None
    source: str = Field(default="upload", max_length=16)
    email_id: str | None = None
    source_filename: str | None = None
    stored_path: str | None = None
    status: str = Field(default="uploaded", max_length=24)
    raw_text: str | None = None
    error: str | None = None
    parsed_at: datetime | None = None
    created_at: datetime = Field(default_factory=utc_now)


class StatementRead(BaseModel):
    id: int
    account_id: int
    period_start: Optional[date] = None
    period_end: Optional[date] = None
    previous_balance: Optional[float] = None
    closing_balance: Optional[float] = None
    minimum_payment: Optional[float] = None
    due_date: Optional[date] = None
    source: str
    status: str
    source_filename: Optional[str] = None
    error: Optional[str] = None
    parsed_at: Optional[datetime] = None

File: frontend/src/types/statement.ts
export interface ReconcileItem {
  status: 'matched' | 'only_statement' | 'only_ledger' | string
  score: number
  statement_tx_id: number | null
  ledger_tx_id: number | null
  date: string | null
  amount: number
  statement_description: string | null
  ledger_description: string | null
  ledger_source: string | null
}

export interface ReconcileReport {
  statement_id: number
  account_id: number
  matched: number
  only_statement: number
  only_ledger: number
  items: ReconcileItem[]
  applied: boolean
  merged?: number
}

export interface Statement {
  id: number
  account_id: number
  period_start: string | null
  period_end: string | null
  previous_balance: number | null
  closing_balance: number | null
  minimum_payment: number | null
  due_date: string | null
  source: string
  status: string
  source_filename: string | null
  error: string | null
  parsed_at: string | null
}
